### ISSUE TITLE
Request: Extract config loading into a separate module

### ISSUE THREAD

**Issue by SimenB** [eslint/eslint#7339]
**What version are you using?**
3.7.1

**What did you do?**
Checked the source code to figure out what module you used to load config for a file (walking up the file tree), verify it, apply defaults, etc

**What happened?**
Found out that the code for it is part of this repo (mostly https://github.com/eslint/eslint/tree/master/lib/config and https://github.com/eslint/eslint/tree/master/lib/config.js)

**What did you expect to happen?**
That it was at least somewhat an external module which could be reused in other projects

My request would be to extract it so that I could reuse the logic in a project I'm working on (https://github.com/SimenB/stylint).

The API I'd roughly want would be to provide a file path, and have the module return a normalized config object from looking at some file for config (`.eslintrc` here), apply options provided there (in the current eslint format), walk up the tree until root of FS or `root: true` (maybe checking package.json along the way), making sure it's normalized. Then return the whole thing.
An awesome addition would be to actually apply defaults before invoking the rule, instead of making the rules themselves handle defaults. Not sure how plausible that is though (considering you haven't done it yet)

I'd be willing to work on this, and provide an initial implementation, but if the extracted code could still be part of the ESLint org, that would be awesome! And if you're not interested, I'll extract a subset of what I need, instead of "wasting time" making it into a generic module

I tried asking on [Gitter](https://gitter.im/eslint/eslint?at=57f8e381a7c618cd26f3f20f) but didn't receive any response.

**Comment by vitorbal** [comment]
I think this might be a duplicate of https://github.com/eslint/eslint/issues/6651.

I'm guessing the comments there still apply, specifically:

> I think it's part of our long-long-term vision, but I'm not really sure we are ready for something like that right now. Also in some cases it will be problematic to separate items you listed above, since, for example, code that deals with configuration (extensions, cascading, extending, etc.) rely on our code that traverses directories.

> Agree with @ilyavolodin -- while we have talked about it as a potential way to go in the future, we are definitely not the point where it makes sense to consider such a change.


However since you're willing to put it the work for this, maybe @nzakas and @ilyavolodin want to re-consider their opinion.

**Comment by SimenB** [comment]
I'd also love to get `glob-util` as it's own thing, as I'm messing about getting a bunch of files to lint right now 😄 

Could go crazy with stuff like `@eslint/glob-util`,  `@eslint/file-util`, `@eslint/config` etc.

**Comment by ilyavolodin** [comment]
Moving all of those bits out, while a good idea, can turn into a maintenance nightmare, if not done correctly. I think it might benefit a lot of people if we were to move it out (for example, stylelint could probably consume `config` package wholesale, same with `sass-lint`) and we might get more contributions for those, but it would definitely make our release process a lot more complicated. We already have trouble with Espree from time to time.
To summarize, I think it's a good idea, just need to be very careful about how it's implemented.

**Comment by SimenB** [comment]
I agree your release process would get more complicated, but what about starting "small", like the globbing I linked to above, then see how it goes? That's not logic that often changes, and should be a stable code base (at least for a JS project). 

Config is more complicated, and I would think it's also code that changes more often that traversing for files

**Comment by platinumazure** [comment]
I'm in favor of trying it out but will defer to the TSC on how much of a headache it would add to the release.

Personally I think we've made great strides already in making the release more reliable (the most recent espree issue was more of a code problem than a release problem, I think). That said, I know that the TSC (especially nzakas) puts in a lot of work behind the scenes to keep things smooth here.

**Comment by SimenB** [comment]
Ok, I'll clean up the "for" then, so it can be looked at as an example

**Comment by SimenB** [comment]
#7345

**Comment by nzakas** [comment]
I'm not comfortable moving a lot of the hidden pieces of ESLint into separate packages directly. Most of the files had very little thought put into their interface, so they're not really appropriate as standalone packages. `config.js` itself is a complete mess that I've been wanting to get rid of for some time, so I definitely wouldn't want to expose that as an interface we have to maintain.

So I would say this: I'm open to splitting out the config logic so long as there is some up front design discussion around what a proper API would look like. @SimenB I'm guessing you have ideas about that since you'd be the primary consumer.

**Comment by SimenB** [comment]
I haven't looked too closely into the actual API of those files, or given much thought to what I'd want out of it.

I'll give it a closer look later today, or at least this weekend, and come back to you with a somewhat structured proposal for an API and feature set 😄

**Comment by SimenB** [comment]
I have looked into it somewhat now, and the current API surface is huge! Mostly because "everything" is exported (to be tested I presume). But all of that wouldn't be part of the public API of an extracted module, so no issue, I suppose.

What I would like as a public API boils down to (with the goal of this being used by linters and its like, not "generic configuration") a single function, `loadConfigForFile(filePath, prefix);`. Some points about it:

- `prefix` (horrible name) is to be able to lookup correct files and options, e.g. `.prefixrc`, `.prefixrc.yaml`, `prefixConfig` in `package.json`, prefix for plugins or configs, etc..
- This would take `filePath`, get its dir, and load config file in that dir, walking up the FS until `root: true` in a file or root of FS is reached, then deep merge it recursively into a single object
 - Would include jumping into `node_modules` to load stuff
   - I suppose we'd also have to pass in some sort of `extendable` array for the config to figure out what parts should be loaded from `node_modules`. `['extends', 'plugins']` for eslint
- We must somehow inject some custom mapping functions for e.g. eslint's `global` and `envs` handling, and other options? Some of it can be special cased generically (such as `rules`), but `global`, `parserOptions` etc is very specific to eslint
 - It would include normalization of rules (always into arrays, with a string severity, for instance)
- I don't know how inline config would be handled (or how it's handled now), but I suppose the `config` instance have some method that can be invoked to change stuff as it walks the AST of a file.

Of course as I write this I feel that it might make sense to create a `ConfigLoader` separately for a bunch of these options (`prefix`, custom mapping functions, what parts of configs should be extendable from external modules). It'd be kinda like an `Ignore` is created for `glob-utils`, then passed in. Mainly to avoid overloading a single functions with too many arguments.

I can try to create a PoC of this if you're interested (and some more hashing out has been done). I just hope it won't be sitting there without feedback, getting stale as other stuff goes on (the tiny #7345 already has a bunch of conflicts after just a few days, and zero feedback).

(Future API I'd like is for rules to be able to be able to get defaults outside of calling the create function of rules, so that create is called with the full config, defaults already figured out. This would be great for documentation purposes, as defaults can then be gotten programmatically. Stretch goal, though!)

**Comment by SimenB** [comment]
Another future API I'd like is something like strict mode, allowing either a severity, ot array of severity and a single object. While this might make people write more in the config, I don't think forcing people to look up docs to understand what the config does is any good.

E.g. `max-len: ["error", 80, 4]`. The `80` on its own is fine, but what does the 4 mean? If we instead enforced `max-len: ["error", { lineLength: 80, tabWdth: 4 }]` or something, it's immediately obvious what it does, no need to look up docs (or dig in code).

I suppose single numeric config could be added as an exception (`indent: ["error", 2]` and `max-len: ["error", 80]` is readable enough), but the point remains

**Comment by not-an-aardvark** [comment]
I agree with you that the `max-len` schema is a bit confusing, but I'm not I agree that the CLI core should regulate that sort of thing. It seems like that's just a matter of designing the API well, but that is subjective enough that I think it would be difficult to do automatically.

**Comment by SimenB** [comment]
Yeah, strict mode (or whatever it's potentially called) should be optional, as I think it's too late for it in eslint's case 😄 

And of course, just in the case it's extracted out (which is the point of this issue).

**Comment by nzakas** [comment]
@SimenB sorry your PR hasn't had much activity - this project is incredibly active and everyone is working in their spare time trying to do their best, so less controversial things tend to get prioritized over others so we can keep the project moving forward. You also flagged your PR as "WIP" and didn't give much context 

I'm a bit concerned about the scope of the changes you keep introducing to this thread, so I'd like to suggest some small, intermediate steps instead. Rather than thinking about how to split out pieces of the ESLint project into independent pieces, let's focus on what a `ConfigLoader` could look like inside of the project as a refactoring plan. If we can figure that out, then we can go the next step and see what it would take to abstract out things that are specific to ESLint.

I'd also like to ask that we keep the discussion on this issue narrowly focused to that one goal. Other small changes should be handled on separate issues. I'd also recommend not submitting a PR first, but rather doing some due diligence and proposing an API that might make sense.

**Comment by SimenB** [comment]
Sure! That PR was good to go I suppose (I had done the code changes that was needed), but still WIP because the name of the module was kinda weird, and the code base should be in this org. Other than that it was perfectly mergable. Should I open a separate issue for the file util/glob stuff? It's a small isolated piece, much smaller in scope than the whole config thing.
Using that extracted module helps me fix a bug in another project, so extracting that now would be great (I can just use the existing extraction of course, if you wanna wait).

But ok, `ConfigLoader`! I did bring up a few points above here. I guess it's not in the form you wanted it since you didn't comment on it, any pointers for what you're asking? Code samples, API suggestions, feature discussion, etc.?

I really appreciate this not getting shut down btw, and I understand you're very busy, with this not being something that will improve eslint directly (although it might make the code base easier to work with 😄 )

**Comment by nzakas** [comment]
I'm not sure I really want to go down the path of separating out file/glob stuff at this point. Those are both so small and with some pretty poorly thought-out APIs that I don't know that it's worth the trouble to explore splitting those out too much. You can feel free to open an issue if you feel strongly, I'm just less convinced that the overhead of maintaining those separately is as useful as potentially maintaining a config loader separately.

For this issue, what I think would be easiest to put together design-wise is to propose an interface (maybe use TypeScript or Flow notation so it's code rather than words), and then provide some further detail as bullet points. Then, give some example code about how you'd see it being used within ESLint itself and explain what code we could get rid of (`config.js`?) or where various bits of functionality would move.

For instance, I'd really like to get all config-related stuff into the `lib/config` directory, so getting rid of `config.js` and adding `lib/config/config-loader.js` would be a good direction. Are there further pieces of `config.js` that can be split out into `config-file.js` and `config-ops.js` (that was the first refactoring I did)? Or would it make sense to add another utility file for some of what's in `config.js`?

**Comment by kaicataldo** [comment]
Looks like a lot of discussion has taken place - where do we currently stand with this issue?

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
